lib/concurrent/configuration.rb

```diff
@@ -1,16 +1,21 @@
 require 'thread'
-require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/executors'
 require 'concurrent/concern/deprecation'
 require 'concurrent/concern/logging'
-require 'concurrent/utility/at_exit'
-require 'concurrent/utility/processor_counter'
 
 module Concurrent
   extend Concern::Logging
   extend Concern::Deprecation
 
+  autoload :AtExit,             'concurrent/utility/at_exit'
+  autoload :AtomicReference,    'concurrent/atomic/atomic_reference'
+  autoload :Delay,              'concurrent/delay'
+  autoload :ProcessorCounter,   'concurrent/utility/processor_counter'
+  autoload :TimerSet,           'concurrent/executor/timer_set'
+  autoload :ImmediateExecutor,  'concurrent/executor/immediate_executor'
```

Comment:
That's the question at hand. None of the changes in this PR are necessary, just a good opportunity for discussion.

The issue is that we create a lot of circular references to `concurrent/configuration`. So far this hasn't been a problem but it's a source of concern. During my #386 experimentation I noticed that `thread_safe` uses `autoload` so I thought I'd give it a try. All the tests pass and the interpreter warnings have gone away.